src/Stack/MetricsCollector.php

```diff
@@ -0,0 +1,229 @@
+<?php
+namespace Stack;
+
+class MetricsRegistry
+{
+    public function __construct($metrics = [])
+    {
+        $storage = new \Prometheus\Storage\APC();
+        $this->registry = new \Prometheus\CollectorRegistry($storage);
+        $this->register_metrics($metrics);
+    }
+
+    public function register_metrics($metrics = [])
+    {
+        foreach ($metrics as $metric => [$kind, $description, $labels]) {
+            [$namespace, $name] = explode('.', $metric);
+
+            switch ($kind) {
+                case 'histogram':
+                    $this->metrics[$metric] = $this->registry->registerHistogram(
+                        $namespace,
+                        $name,
+                        $description,
+                        $labels
+                    );
+                    break;
+
+                case 'counter':
+                    $this->metrics[$metric] = $this->registry->registerCounter(
+                        $namespace,
+                        $name,
+                        $description,
+                        $labels
+                    );
+                    break;
+
+                case 'gauge':
+                    $this->metrics[$metric] = $this->registry->registerGauge(
+                        $namespace,
+                        $name,
+                        $description,
+                        $labels
+                    );
+                    break;
+            }
+        }
+    }
+
+    public function get_gauge($metric)
+    {
+        return $this->metrics[$metric];
+    }
+
+    public function get_counter($metric)
+    {
+        return $this->metrics[$metric];
+    }
+
+    public function get_histogram($metric)
+    {
+        return $this->metrics[$metric];
+    }
+
+    public function render()
+    {
+        $renderer = new \Prometheus\RenderTextFormat();
+        $result = $renderer->render($this->registry->getMetricFamilySamples());
+
+        header('Content-type: ' . \Prometheus\RenderTextFormat::MIME_TYPE);
+
+        echo $result;
+        exit();
+    }
+}
+
+class MetricsCollector
+{
+    const REQUEST_TYPES = array( 'cron', 'api', 'admin-ajax', 'admin', 'search', 'frontpage', 'singular', 'archive', 'other' );
+
+    private $registry;
+    private $metrics;
+    private $wpdb_stats;
+
+    public function __construct()
+    {
+        $this->metrics = new MetricsRegistry(
+            array(
+                'wp.requests'              => array('counter',   'Number of requests',                                 ['request_type']),
+                'wp.page_generation_time'  => array('histogram', 'Page generation time, in seconds',                   ['request_type']),
```

Comment:
It's easy to have a type in the metric's name. Let's move them to constants.

Reply:
Typo you mean?  
So something like STACK_METRICS_WP_REQUESTS = 'wp.requests' ?